### PR TITLE
Command support

### PR DESCRIPTION
**Pull request by gregquinn2001**
JIRA 1203. Adds a javascript function at root level for submitting commands to Flux Notes. javascript function can be called from an iOS app or Android app with Flux Notes embedded in a webview.

To test this, open console and insert one of the following commands. Note that the insert structured phrase ones require a note to do much (either in editor or point of care).

Commands are as follows:
flux_command('insert-structured-phrase', {phrase:'toxicity', fields: [{name:'adverseEvent', value: 'nausea'}, {name:'grade', value: 'grade 2'}]})
flux_command('insert-structured-phrase', {phrase:'toxicity', fields: [{name:'adverseEvent', value: 'anemia'}, {name:'grade', value: 'grade 3'},{name:'attribution', value:'Treatment'}]})

flux_command('insert-structured-phrase', {phrase:'disease status', fields: [{name:'status', value: 'Stable'}, {name:'reasons', value: ['Physical exam', 'Symptoms']}]})

flux_command('insert-structured-phrase', {phrase:'enrollment', fields: [{name:'title', value:'PATINA'}]})

flux_command('navigate_targeted_data_panel', {section:'Medications'})
flux_command('navigate_targeted_data_panel', {section:'Timeline'})

Any section in targeted data panel can be navigated to (using short or long name for section).

inserts into open note or point of care
returns error message from flux_command so voice or whatever external app is sending commands knows it was unsuccessful (when result is not null it is a string which is an error message for user) and why and can let user know if appropriate.
Multiple toxicities can be inserted while the other 3 replace any previous values.

_Pull requests into Flux require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:._

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome
- [ ] Manually tested in IE

**Documentation**

- [x] This pull request describes why these changes were made
- [ ] Recognizes any potential shortcomings/bugs in the description 
- [ ] Cheat sheet is updated
- [ ] Demo script is updated 
- [ ] Documentation on Wiki has been updated 
- [ ] Additional technical components and libraries are documented and added to wiki as needed

**NoteParser**

- [ ] Note parser has been updated if structured phrases change

**Code Quality**

- [ ] 4-space indents - convert any tabs to spaces
- [ ] Use public class field syntax for binding functions - ex. handleChange = () => { ... };
- [ ] Code is commented

**Tests**

- [ ] Existing tests passed
- [ ] Added Enzyme-UI tests for slim mode 
- [ ] Added Enzyme-UI tests for full mode
- [ ] Added backend tests for new functionality added
- [ ] Added note parser examples to test new functionality

## Reviewer 1: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

## Reviewer 2: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: src/notes/FillPlaceholder.jsx [hunk 1]
```diff
@@ -67,6 +67,38 @@ export default class FillPlaceholder extends Component {
...
     }
 
+    fillFromData = (data) => {
+        const { placeholder } = this.props;
+        let errorToReturn = null, error; // return first error only
+
+        let entryIndex = 0;
+        if (placeholder.multiplicity === 'many') {
+            let lastEntryIndex = placeholder.entryShortcuts.length - 1;
+            const valFirstField = placeholder.getAttributeValue(data.fields[0].name, lastEntryIndex);
```
Comment (danlee1025) [inline src/notes/FillPlaceholder.jsx]: [DL1]  The logic here is looking at the first field of the last entry to determine if it should add a new entry or fill out the existing entry.  Should it look at all the fields and see if any of them are already filled out?

**Comment by danlee1025** [comment]
Looks very cool Greg!  I didn't even know something like this was possible.

**Comment by danlee1025** [comment]
[DL2]  When I enter `flux_command('insert-structured-phrase', {phrase: 'staging', fields: [{name:'T', value:'T0'}]})`  I get the result below.  Should it return an error?
screen shot 2018-08-16 at 9 12 19 am

**Comment by gregquinn2001** [comment]
For DL1: Good point Dan. The issue I have is defaulting as we discussed. I'll add the ability to tag a shortcut with the source. Once a shortcut has a source, it's filled in and won't be used again by a command.

**Comment by gregquinn2001** [comment]
[All the ways to create a shortcut in Flux Notes.docx](https://github.com/FluxNotes/flux/files/2304130/All.the.ways.to.create.a.shortcut.in.Flux.Notes.docx)

**Comment by gregquinn2001** [comment]
DL1 is resolved. We know set a source attribute on every shortcut to track how it was created. If the last toxicity entry has a source attribute, a new one is added now. An entry gets a source when the first value is set based on how it is set.

**Comment by gregquinn2001** [comment]
also rebased

**Comment by gregquinn2001** [comment]
For DL2, I tried this and got the following:

flux_command('insert-structured-phrase', {phrase: 'staging', fields: [{name:'T', value:'T0'}]})
null

so I get null return result which is correct...

**Comment by nicoleng12** [comment]
This is really cool! Nice Job Greg. I tested in IE and in chrome. The backend tests pass and walking through the demo script everything looks good.

I noticed one thing with the navigating in the targeted data panel:

[NN1] When I scroll in the targeted data panel (either via mini map or scroll bar) and then enter the command to jump to the medications section, the targeted data panel goes to the "Labs" section, not the "Medications" section. In general, changing the view in the targeted data panel and then entering the command to navigate to a section doesn't go to the section you would expect

**Comment by gregquinn2001** [comment]
Actually on NN1, let's create a JIRA. The command support was not required to support moving sections initially. The voice app will only create structured phrases using the command support for now. I'll create a JIRA now.

**Comment by gregquinn2001** [comment]
JIRA created. no rebase needed. Please merge!

**Pull request merged**